src/session.c

```diff
@@ -431,8 +473,15 @@ static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
             };
             ret = pw_cb(cb_pin, sizeof(cb_pin), &cb_pin_len, params, pw_cbarg);
             if (ret != RET_OSSL_OK) {
-                ret = CKR_GENERAL_ERROR;
-                goto done;
+                /* this error can mean anything from the user canceling the prompt to no UI method provided
+                 * Fall back to our prompt here */
```

Comment:
if we did have a callback I do not think we should repeat the prompting,
we should prompt on our own if no callback is present, gated behind a config option that enables this behavior.

Reply:
As described in this comment, from store, we are always getting a callback, but if the calling application does not provide the UI, it just fails. I think we are getting the pointer to `ossl_pw_get_passphrase()` function from `crypto/passphrase.c` unconditionally. And based on what application provided, it works of fails with `No password method specified`:
https://github.com/openssl/openssl/blob/cf6342bc024868f5a55f2225f2e083415fb1329a/crypto/passphrase.c#L275